### ISSUE TITLE
Zed Taking 185 GB of Ram on my Macbook due to Potential Memory Leak Issues

### ISSUE THREAD

**Issue by Hchanni** [zed-industries/zed#17937]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

Hitting some insane amounts of memory leak issues where it's saying Zed is taking 140 GB of RAM. 

### Environment

Apple M1 Pro 
Memory - 32 GB
MacOs - Sonoma - 14.6.1

### If applicable, add mockups / screenshots to help explain present your vision of the feature

![image (1)](https://github.com/user-attachments/assets/add73a73-3417-4e33-933e-89bce4a6aa99)

### If applicable, attach your Zed.log file to this issue.

<details><summary>Zed.log</summary><pre>

</pre></details>

**Comment by JadoJodo** [comment]
Same on Sonoma 14.6.1.

I've been able to recreate it by having a window open that has ~ in scope. I haven't dug into it much, but my suspicion is that it's due to the macOS folders that come with access to ~ (Drives, Desktop, etc).

**Comment by brandondrew** [comment]
My computer has 16 GB or RAM, but Zed tried to grab far more after I typed `zed .` in my home directory and tried comparing files across multiple projects.

<img width="539" alt="image" src="https://github.com/user-attachments/assets/1aecf448-71ba-4a1c-a88b-8c47d5544cfc">

I don't know whether the problem was because of a search I did or because of having the file pane open.  If I remember correctly, I had killed the search and given up on getting any results, and the problem came later.  When I saw the memory problem I had forgotten about the search and initially assumed it was because of the file pane in a large directory.

After force-quitting Zed, I opened it again to recover what I had before.  I closed the file pane, and didn't notice the memory consumption going out of control before I closed that window.  However, this is not conclusive, since I was more concerned with continuing my work than empirically testing whether the file pane being closed would allow me to keep working with the window open for a prolonged time.

Actually, there's a third possibility.  I saw several downloads being performed mentioned in the status bar.  I don't know if that's connected, but it was going on while the memory was out of control.


```
Zed: v0.155.2 (Zed)
OS: macOS 14.7.0
Memory: 16 GiB
Architecture: aarch64
```

**Comment by brandondrew** [comment]
https://gist.github.com/brandondrew/f3d496b34a1bbccb56dd7f0ef5cf2bd6

**Comment by pchalasani** [comment]
> a window open that has ~ in scope.

can you elaborate on this? Didn't know what you meant. I had the same problem yesterday, zed blew up to 85GB on my M1 Pro Max.

**Comment by JadoJodo** [comment]
@pchalasani 

Since macOS mounts various services (e.g., iCloud, via Documents, Desktop, etc.), as well as the Library folder (which also has tendrils that go all over the place) in ~ (Home folder), my theory is that opening that folder causes Zed to dive into infinite rabbit-holes.

**Comment by dimitrisnl** [comment]
I can confirm that having a window open with `~` in scope does it. The moment this happened, it went from ~400MB to 29GB, and kept rising. 

![CleanShot 2024-10-11 at 21 40 24@2x](https://github.com/user-attachments/assets/7f3e1999-78d9-4e69-b129-51c858a4386f)

**Comment by pchalasani** [comment]
> having a window open with `~` in scope

Trying to nail down precisely what this means, in case it helps others too. Do you mean a terminal/console window within zed where the current dir is ~, or something else? I don't recall having such a terminal window, but maybe I did temporarily.

**Comment by JadoJodo** [comment]
@pchalasani 

If you open the macOS home folder in Zed, this issue will occur.

**Comment by pchalasani** [comment]
> @pchalasani 
> 
> 
> 
> If you open the macOS home folder in Zed, this issue will occur. 

Thanks for clarifying. But I've never actually opened the home folder like this, which makes this puzzling.

**Comment by brandondrew** [comment]
@pchalasani maybe you had a symlink to `~`, or you had something else very large open (or a symlink to something else very large in your project).  `~` may just be the most common example, but not necessarily the only directory with the necessary size (or other properties) to cause the problem.

**Comment by brandondrew** [comment]
<img width="539" alt="image" src="https://github.com/user-attachments/assets/4e2ddd21-61f2-4908-a982-82505e2c5c83">

I had several projects open when this happened.  As far as I know none of them had any symlinks to `~`, though.  If I find anything like that in one of them, I'll report back here...

I closed them down one by one, and after closing the last one, the memory consumption gradually dropped to normal levels.  They were open for days, though, across a couple of upgrades of Zed, so there were no signs of this being caused simply by having them open in Zed.

**Comment by felixsanz** [comment]
not so much, but still a problem

<img width="277" alt="image" src="https://github.com/user-attachments/assets/b628017b-96f1-4114-b602-fad6525b24f4">

**Comment by notpeter** [comment]
Beyond the obvious example of `zed ~`, I'm curious what conditions / contents of the workspace directories trigger this behavior.  One example observed internally was using "Search  project with 33x5GB JSON temporary files in the workspace (not covered by .gitignore).

Does anyone have an example of an open source repo that triggers this ballooning?

**Comment by notpeter** [comment]
There are multiple potential causes of runaway memory consumption:

Zed project search: 
- https://github.com/zed-industries/zed/issues/20970

Ansible Extension, caused by tree-sitter-yaml (fixed, but still no public repro)
- https://github.com/kartikvashistha/zed-ansible/issues/2

TOML Extension on macOS, caused by Taplo language server:
- https://github.com/zed-industries/zed/issues/13190

And surely others.
To help narrow down whether your issue is a subprocess or Zed/Zed Extensions, you may wish to keep an eye on the Memory Tab of Activity Monitor and select "View->All Processes Hierarchically.
<img width="788" alt="Screenshot 2024-11-21 at 1 23 25" src="https://github.com/user-attachments/assets/b978e0ec-2be6-44d2-a8b4-f77c6ce34364">

**Comment by brandondrew** [comment]
@notpeter so asking why "the obvious example of `zed ~`" is actually obvious is off topic?

**Comment by notpeter** [comment]
No, the example isn't off-topic, the request 20 comments deep in a thread about zed consuming all memory to discuss Zed's project scanning behavior is off-topic.  In my experience, `zed ~` has no issues scanning the 414GB of 3.6million files in my home directory. It takes 8 minutes, but amazingly Zed is responsive throughout this time, and memory usage never exceeded 3.5GB. 

Zed is a code editor not a file explorer -- features like "cmd-p" for open project files cannot work if you don't have an index of files. In my home directory, VSCode does not pre-index and if I try and open a file this way it will happily show a spinner for multiple minutes without showing any results.  Subsequent searches will take the same multiple minutes to complete.

For me, one action that triggered memory ballooning was using Project Search in my home directory (Zed has to scan through 3.6million and open a subset of them and then linerally scan through them). When there are large text files (e.g. big JSON, big XML) Zed exhibits pathological behavior which is why I opened https://github.com/zed-industries/zed/issues/20970

If you'd like to have a discussion about Zed architecture you are free to open a GitHub discussion.  In the meantime, I am going to stay focused on creating test cases that can reproduce Zed to use tons of memory.

I believe there are multiple causes of this issue -- if you experience this again, can you open Activity Monitor to the memory tab with "Show Processes Hierarchically" to see where the memory is used, specifically whether it is Zed or it's subprocesses (LSPs, etc) which may help identify the root cause of your issue (and potentially a workaround, like disabling taplo if it's choking on your TOML files).

**Comment by notpeter** [comment]
I'm going to close this as unactionable.  I've [linked](https://github.com/zed-industries/zed/issues/17937#issuecomment-2490468962) to multiple other issues above.

If you are experiencing runaway memory when opening a large folder (e.g. `zed ~`) please take a moment to attempt to bisect which folder(s) might be causing this issue and see if you can come up with exact steps to reproduce (memory balloon on load; on project search; when loading a particular language server) and verify whether it is Zed or subprocesses by opening Activity monitor and selecting Memory, then View->All Processes hierarchically so you can explicitly attribute memory use (e.g. if an LSP is ballooning memory usage, not Zed, the macOS Force Quit Applications shows that memory as being zed -- those are two very different problems to troubleshoot). 

Since no one is going to share a tarball of their homedirectory, we need to have narrower repros of the (likely multiple) root causes. Please feel free to file clean issues if you have steps sufficient to reproduce on a clean machine. Thanks!
